### ISSUE TITLE
Issues with vertex layout when using VK_LAYER_KHRONOS_shader_object

### ISSUE THREAD

**Issue by maksym-pasichnyk** [KhronosGroup/MoltenVK#2051]
When I try use VK_LAYER_KHRONOS_shader_object geomeometry is not visible.
MoltenVK set MTLBufferLayoutStrideDynamic into MTLVertexBufferLayoutDescriptor, and geometry is not drawing. 
When I change line below (first screenshot) and set pVKVB->stride into vbDesc.stride everything working fine and geometry is visible (second screenshot).

<img width="860" alt="Screenshot 2023-10-28 at 02 37 58" src="https://github.com/KhronosGroup/MoltenVK/assets/11374170/4b6c763e-2349-4a46-86b8-b6525f06ee56">
<img width="912" alt="Screenshot 2023-10-28 at 02 42 45" src="https://github.com/KhronosGroup/MoltenVK/assets/11374170/782a8ea1-f6c7-4729-86f9-3018c01ffff3">

**Comment by cdavis5e** [comment]
That would indicate that you (or the layer you're using, since `VK_EXT_shader_object` also supports this) set vertex input strides to be a dynamic state, but you're not calling `vkCmdBindVertexBuffers2()`. According to [the Vulkan spec](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap10.html#VkDynamicState):

> * `VK_DYNAMIC_STATE_VERTEX_INPUT_BINDING_STRIDE` specifies that the `stride` state in [VkVertexInputBindingDescription](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap22.html#VkVertexInputBindingDescription) will be ignored and **must** be set dynamically with [vkCmdBindVertexBuffers2](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap22.html#vkCmdBindVertexBuffers2) before any draw call.

In other words, you have to call `vkCmdBindVertexBuffers2()` or the behavior is undefined--it is allowed to crash, corrupt your hard disk, make demons fly out of your nose, or even *work just fine* if you don't call `vkCmdBindVertexBuffers2()` in this case. But note that it doesn't have to "work just fine" *everywhere*. The thing about undefined behavior is that it *varies* from implementation to implementation.

What happens when you use the validation layer? Are you calling `vkCmdBindVertexBuffers2()` like you're supposed to? Are you calling `vkCmdSetVertexInput()` instead? The latter won't work with MoltenVK because it doesn't implement `VK_EXT_vertex_input_dynamic_state`.

**Comment by maksym-pasichnyk** [comment]
@cdavis5e Yes, I'm calling vkCmdBindVertexBuffers2 and specify strides, but they are ignored in MoltenVK  because IsVtxStrideState is true. Nothing happens when I use validation layer, no errors, no warnings

**Comment by richard-lunarg** [comment]
Which version of macOS are you running? You need Metal 3.1 for VK_DYNAMIC_STATE_VERTEX_INPUT_BINDING_STRIDE. I'm not sure the extension should actually be exposed unless Metal 3.1 is available and the full functionality is available.

**Comment by cdavis5e** [comment]
> @cdavis5e Yes, I'm calling vkCmdBindVertexBuffers2 and specify strides, but they are ignored in MoltenVK because IsVtxStrideStatic is false. Nothing happens when I use validation layer, no errors, no warnings

??

You're not making sense. The code you cited sets up the Metal vertex input descriptor for the pipeline object. It sets the ***STATIC*** vertex input state--you know, what would be used if you were using pipeline objects without `VK_EXT_extended_dynamic_state` or `VK_EXT_shader_object`. `MTLBufferLayoutStrideDynamic` is used in Metal pipeline vertex input descriptors to mark dynamic vertex input strides.

...

Are you calling `vkCmdBindVertexBuffers2()`... or `vkCmdBindVertexBuffers2EXT`() (with suffix)? You actually need to call the latter on MoltenVK, because MoltenVK doesn't support Vulkan 1.3.

**Comment by JagexDark** [comment]
Also ran into this and found that force disabling `VK_EXT_extended_dynamic_state` allows `VK_LAYER_KHRONOS_shader_object` to function correctly with `vkCmdSetVertexInputEXT` and `vkCmdBindVertexBuffers`.

I'm inclined to agree that `VK_EXT_extended_dynamic_state` should not be exposed without full support, or alternatively logic could be added to the shader object layer to detect this situation (and also still allow the other parts of `VK_EXT_extended_dynamic_state` to be used)
